### ISSUE TITLE
ability to run lua program line-by-line / pause execution and return to caller

### ISSUE THREAD

**Issue by jvantuyl** [rvirding/luerl#137]
I'm trying to write a distributed consensus based sandbox.  This would be absolutely perfect if I could somehow step through the program and do consensus on the VM state.  Unfortunately, I can't figure out any way to pause execution.  Ideally, the requested feature would work something like:

    InterpState = luerl:call_func(Code, {enable_breakpoints, [{explicit, true}, {reductions, 100}])
    {Status, State} = luerl:run(InterpState)

Or something like the above.  Basically some way to say "run like normal but break out and give me a status code and a paused interpreter state (that's opaque but can be serialized) when something explicitly triggers a breakpoint or every 100 'reductions'".

I recognize this is a big ask and I suspect the answer will be "not likely to happen", but I thought I'd at least ask.  ^_^

**Comment by jvantuyl** [comment]
Wait... I just ran across `luerl_sandbox.erl`.  Is that half of this?  Is it possible to explicitly pause the VM anywhere?

**Comment by linux2647** [comment]
I have a similar need, and it seems like the tracing functions could give you what you need.  Instead of every 100 reductions, it calls the tracing function (and pauses the VM execution) on every line, function call, and return.  I haven't used it yet, but it seems promising.  See this [commit](https://github.com/rvirding/luerl/commit/69b1107e481f6172188ab929c41e30f949d56438) for more (I don't see any other documentation on it).  It was also included in the [v1.0 release](https://github.com/rvirding/luerl/releases/tag/v1.0)

**Comment by rvirding** [comment]
Yes, one way to do this would be to use the tracing function. It is a @linux2647 mentioned call every line, function and return with the traced op and the current luerl state. It returns the potentially updated luerl state. Note be very careful if you modify the state as this is what is then used in the interpreter. There is `trace_data` field in the luerl state which is reserved for the tracer and is not used by anything else. Here is a very simple example file (unfortunately not in the in the commit, sorry about that) [trace_test.erl.zip](https://github.com/rvirding/luerl/files/7741233/trace_test.erl.zip). If you can't read it I will just paste the content.

**Comment by jvantuyl** [comment]
Thanks for this.

Hmmmm... I'm not entirely sure if tracing will work.  Ideally I'd have something more like a continuation.

Sorry if my use-case isn't super clear.  What I'm trying to do is build a sandbox that can be run on multiple hosts and then use consensus to provide fault-tolerance.  I figured I could just run up to the point that it does some kind of IO and that would be my sync point between the various nodes.

What I need is essentially:
* A way to start an interpreter from scratch (this isn't a problem today).
* A way to have a call in the interpreter bail out and give me a ball of state.
* A way to resume that ball of state providing a return value.
* A way to serialize that state and send it between nodes.

I'm not 100% sure if the last one works or not.  If there are anonymous functions in there, I just don't know what it would do.  (I figured I'd just try and see what explodes.)

By way of example, let's say I wanted to have a sandboxed program that runs the following script:
```lua
function heartbeat ()
  while true
    -- sleep 5 seconds
    sandbox.sleep(5000) 
    -- log a heartbeat
    sandbox.print("heartbeat")
  end
end
```

Ideally, it'd look like this from the Erlang side:
```erlang
state0 = luerl:init()
-- agree on state via consensus
-- agree on startup via consensus
{state1, info_about_continuation_call}  = luerl:do(heartbeat_script)
-- agree on new state via consensus
-- agree on when to continue execution and on return value of nil
state2 = luerl:continue(nil, state1}
-- loop on this until the program is done
```

And presumably the sandbox library code would look something like this:
```lua
sandbox = {}

function sandbox.sleep(time)
  return magic_luerl_internals.escape_hatch("sleep:" .. type(time) .. "=" .. time)
end
```

The idea is to be able to break out at any point, serialize what I've got, send it somewhere else, and then be able to resume it.

Tracing seems to be kind of inside-out from what I need.  I could probably use it to capture state, but I'm not entirely sure how I'd take that serialized state and send it somewhere else.

I know that this is probably not an expected use-case.  Given Erlang's functionalness and Lua being such a widely understood language, it would be really great if I could make this work.  I've already tried this with Python and it didn't go so well.

If this isn't something that you want to implement, could I possibly spike something out and see what you think of it?  Luerl has solved all of the hard parts of what I need to do.  I just need the right control surface.
